### PR TITLE
Added Repo.embedded_load/3 function

### PR DESCRIPTION
**Pull request by narrowtux**
as described in #3269

### CODE REVIEW

**Comment by josevalim** [comment]
Oh, interesting. Given this function doesn't use the adapter, maybe it doesn't belong on Repo but elsewhere. The question is where? `Ecto.Schema`? `Ecto`?

**Comment by josevalim** [comment]
Btw, really good job. The question is only where to place the code.

**Comment by narrowtux** [comment]
Yeah that's true. I think it belongs in Schema then.

**Comment by josevalim** [comment]
@wojtekmach @michalmuskala any thoughts?

**Comment by wojtekmach** [comment]
> maybe it doesn't belong on Repo but elsewhere. The question is where? Ecto.Schema? Ecto?

Yeah, neither option seems optimal to me. Given embedded_load/3 is implemented and documented almost identically to Repo.load/2, I think I'd keep it on the Repo regardless of the caveat you mentioned (which I totally agree with). I think putting it on `Ecto.Schema` would be fine too but then let's mention both functions in their respective docs imho.

**Comment by michalmuskala** [comment]
Yes, I agree. We'll need at least a lot of cross-linking between them since they are related. I think there's another question and that is loading data that is not from database into the schema. Since, as far as I remember, for embeds load and cast is exactly the same this is generally already possible with `Changeset.cast` followed by `Changeset.apply_changes`. What should be the interaction there? It seems to me this might be a bit confusing as to what function you're supposed to use in what situation.

**Comment by narrowtux** [comment]
> for embeds load and cast is exactly the same this is generally already possible with Changeset.cast followed by Changeset.apply_changes. What should be the interaction there? It seems to me this might be a bit confusing as to what function you're supposed to use in what situation.

One advantage over the changeset calls is that you don't need to repeat the list of fields.

The changeset approach offers more flexibility of course.

**Comment by josevalim** [comment]
>  I think there's another question and that is loading data that is not from database into the schema.

Good call. In those cases we may not have a repo, so putting it on a repo is a no-go. I looked at Ecto.Schema and it only has the schema definition functionality, functions that work on the schema at runtime are all in Ecto, so I propose to add it as `Ecto.embedded_load` and work on cross-linking them in the docs.

Can you please amend it acccordingly @narrowtux ? Thanks everyone!

**Comment by narrowtux** [comment]
Okay

**Comment by narrowtux** [comment]
@josevalim I moved the embedded_load/3 function to the Ecto module, and moved the test to `embedded_test.exs`.

File: lib/ecto.ex [hunk 1]
```diff
@@ -601,4 +601,40 @@ defmodule Ecto do
       true
     end
   end
+
+
```
Comment (josevalim) [inline lib/ecto.ex:605]:
```suggestion
```

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Comment by josevalim** [comment]
It feels like Ecto.embedded_dump/3 would really complement this feature, but this is good enough for getting started. Thanks @narrowtux !

**Comment by narrowtux** [comment]
Yeah, an `embedded_dump` makes sense! It probably also works with the `Ecto.Repo.Schema.do_load` function.

**Comment by narrowtux** [comment]
Also, I noticed that this function doesn't load the data into recursive schemas. I'll continue to work on this a bit

**Pull request merged**